src/malli/design_draft.clj

```diff
@@ -0,0 +1,96 @@
+(ns malli.design-draft
+  (:import (clojure.lang IPersistentMap)))
+
+
+;;; malli schemas
+
+;; Core schemas are not namespace qualified keywords to make using them user
+;; friendly.
+
+;; Schemas are written using Hiccup style vectors, where order is significant.
+
+;; Every schema definition can have additional options-map after schema name.
+;; Options-map can have arbitrary amount of data, with some keys having special
+;; meaning, like below, which describes a closed map.
+
+[:map {:closed true, :description "My closed map"}]
+
+;; Schema definitions live in registry. Malli provides a core registry for
+;; all the essential schemas. Registry can be extended on the runtime and it's
+;; possible to have as many registries one needs.
+
+;; The actual registry definitions and format is wip, but could be something
+;; like below.
+
+;; Compiles given definition to Schema
+(declare compile)
+(declare Schema)
+
+(def -map (compile {:id :map
+                    :type IPersistentMap
+                    :title "Map"
+                    :json-schema/type "object"
+                    :valid-fn (fn [data])}))
+
+(def -and (compile {:id :and
+                    ;; Does this make any sense?
+                    :json-schema/op "allOf"}))
+
+;; or create Schema directly
+(def -map (reify Schema '...))
+
+(def core-registry
+  {:map -map
+   :and -and})
+
+;; We could expose core schemas as vars as well for having development time
+;; documentation etc.
+(def ^{:doc "A map"} map :map)
+
+
+;;; TBD List core schemas
+;;; TBD List main functions operating on schemas
+;;; TBD Examples of validation and validation error messages
+;;; TBD Examples of coercion
+;;; TBD Examples of schema format transformation eg. to Json schema
+
+;;; Larger schema example
+
+;; Creating an inlined model inspired from schema.org Person
+(def example1
+  [:schema
+   [:map {:id :my/Thing
+          :title "A Thing"
+          :description
+          :schema.org/type "http://schema.org/ImageObject"}
+    :identifier [:or :uuid :url [:and :string :not-empty]]]
+   [:and {:id :my/ImageObject
+          :schema.org/type "http://schema.org/ImageObject"
+          :json-schema/id "http://my.org/ImageObject"}
+    [:map {:closed true
+           ;; Use keys from another map schemas?
+           :merge-keys [:my/Thing]
```

Comment:
I'm not fond of the flat sequence of keys and values, harder to program with those, kind of like the `& rest` params which is not used anymore with clojure function args to represent options. Instead, I would use a vector of key+value tuples (vectors) for maps.

so, instead:
```clj
[:map :x :int :y :string]
```
this:
```clj
[:map [:x :int] [:y :string]]
```
In clojure, the map is a sequence of entries, which are tuples. I think we should use that too: you can count the keys with `(count childs)`, not `(/ (count childs) 2)` etc.

if the `expand` starts from the top, it can delegate the `expand` to the child schemas, and `:map` knows how to expand it's children.

Reply:
Ok, let's convert to stricter version, even if the we could allow sugar of kv-tuples without wrapping, that would complicate matters.